pkg/xds/server/callbacks/dataplane_lifecycle.go

```diff
@@ -142,6 +145,10 @@ func (d *DataplaneLifecycle) register(
 		return errors.Wrap(err, "could not register proxy passed in kuma-dp run")
 	}
 
+	// We should wait for Cache ExpirationTime to let MeshContext sync the latest data
+	// in which the latter DataplaneSyncTracker callback relies on it to generate the XDS configuration.
+	// This only happens on Universal Dataplane because the k8s Dataplane object is created by the controller.
+	time.Sleep(d.cacheExpirationTime)
```

Comment:
the problem is that this sleep will block
```
	defer info.mtx.Unlock()
```
so ALL dpps that are trying to connect will be waiting

Reply:
IIUC, the variable `info` is corresponding to a unique `proxyKey`.

```golang
info, loaded := d.proxyInfos.LoadOrStore(proxyKey, &proxyInfo{
  proxyType: md.GetProxyType(),
})
```

And the proxyKey is generated by the envoy nodeID:

```golang
proxyId, err := core_xds.ParseProxyIdFromString(request.NodeId())
if err != nil {
  return errors.Wrap(err, "invalid node ID")
}
dpKey := proxyId.ToResourceKey()
```

So, this lock would only block the unique dataplane object rather than all of them right?
Here's the envoy bootstrap configuration that I downloaded, the `dp-echo-1` is the universal dp name that I defined:

```text
node:
  cluster: backend
  id: default.dp-echo-1
  metadata:
    dataplane.admin.address: 127.0.0.1
```

Reply:
I read this wrong, you are right with this one